lib/commands/file-movement.js

```diff
@@ -1,12 +1,243 @@
+ *                              '@com.myapp.bla/RelativePathInContainer/111.png'. The '@' character at the
+ *                              beginning of the argument is mandatory in such case.
+ *                              The relative folder path is ignored if the file is going to be uploaded
+ *                              to the default media folder and only the file name is considered important.
+ * @param {string} base64Data - Base-64 encoded content of the file to be uploaded.
+ */
+async function pushFileToSimulator (device, remotePath, base64Data) {
+  if (_.isArray(base64Data)) {
+    // some clients (ahem) java, send a byte array encoding utf8 characters
+    // instead of a string, which would be infinitely better!
+    base64Data = Buffer.from(base64Data).toString('utf8');
+  }
+  const buf = Buffer.from(base64Data, 'base64');
+
+  if (remotePath.startsWith(CONTAINER_PATH_MARKER)) {
+    const [bundleId, dstPath] = await parseContainerPath(remotePath,
+      async (x) => await getAppContainer(device.udid, x));
+    log.info(`Parsed bundle identifier '${bundleId}' from '${remotePath}'. ` +
+             `Will put the data into '${dstPath}'`);
+    if (!await fs.exists(path.dirname(dstPath))) {
+      log.debug(`The destination folder '${path.dirname(dstPath)}' does not exist. Creating...`);
+      await fs.mkdirp(path.dirname(dstPath));
+    }
+    await fs.writeFile(dstPath, buf, 'binary');
+    return;
+  }
+  const dstFolder = await tempDir.openDir();
+  const dstPath = path.resolve(dstFolder, path.basename(remotePath));
+  try {
+    await fs.writeFile(dstPath, buf, 'binary');
+    await addMedia(device.udid, dstPath);
+  } finally {
+    await fs.rimraf(dstFolder);
+  }
+}
+
+/**
+ * Save the given base64 data chunk as a binary file on the device under test.
+ * ifuse/osxfuse should be installed and configured on the target machine in order
+ * for this function to work properly. Read https://github.com/libimobiledevice/ifuse
+ * and https://github.com/osxfuse/osxfuse/wiki/FAQ for more details.
+ *
+ * @param {Object} device - The device object, which represents the device under test.
+ *                          This object is expected to have the `udid` property containing the
+ *                          valid device ID.
+ * @param {string} remotePath - The remote path on the device. This variable can be prefixed with
+ *                              bundle id, so then the file will be uploaded to the corresponding
+ *                              application container instead of the default media folder, for example
+ *                              '@com.myapp.bla/RelativePathInContainer/111.png'. The '@' character at the
+ *                              beginning of the argument is mandatory in such case.
+ * @param {string} base64Data - Base-64 encoded content of the file to be uploaded.
+ */
+async function pushFileToRealDevice (device, remotePath, base64Data) {
+  await verifyIFusePresence();
```

Comment:
should we do the same base64 transformation here? perhaps, the transformation itself might be moved to appium-support, since it might be useful for Android as well?

Reply:
ah, yes we should. i mean _really_ what we should do is fix the client, because it should be sending a string. i just needed something i could fix faster than the java client :-)

Reply:
I'll take care of it. But there are still many older versions in the wild

Reply:
usually people update server much more frequently than the client